### PR TITLE
Integrate `pre-commit` 

### LINKED ISSUE #124: Add `pre-commit` hook
See: https://pre-commit.com/

### PR DESCRIPTION
**Pull request by alexnguyenn**
Closes #124 

How to run:
* `pip install pre-commit` and then `pre-commit install` 
* To run: `pre-commit run` or just `git commit ...`, the first time running will take a while since it has to clone all the repos specified in `pre-commit-config.yaml`
* If want to run against all files (good for first run to test things out): `pre-commit run -a`.

### CODE REVIEW

**Comment by alexnguyenn** [comment]
@gwwatkin with `pre-commit install`, I am wondering whether or not we need `mypy` and `flake8` under `setup.cfg`. If a contributor don't use `pre-commit`, we can just do the checks via Github Actions once they submit a PR. 

Same case can be made for `isort` and `black`, but it is harder to decide since they are used as formatters. Right now `pre-commit run` runs them with `--check` so they just check if the code meets their specifications before a commit is made.

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+  - repo: https://gitlab.com/pycqa/flake8
+    rev: 4.0.1
+    hooks:
+      - id: flake8
+
+  - repo: https://github.com/pre-commit/mirrors-mypy
+    rev: v0.910
+    hooks:
+      - id: mypy
```
Comment (gwwatkin) [inline .pre-commit-config.yaml:36]: Shouldn't there be a `--ignore-missing-imports` in the args? can't test right now though

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    rev: 5.10.0
+    hooks:
+      - id: isort
+        args: ["--check"]
+
+  - repo: https://github.com/ambv/black
+    rev: 21.10b0
+    hooks:
+      - id: black
+        args: ["--check"]
```
Comment (gwwatkin) [inline .pre-commit-config.yaml:26]: Shouldn't we specify the line length here?

**Comment by gwwatkin** [comment]
> @gwwatkin with `pre-commit install`, I am wondering whether or not we need `mypy` and `flake8` under `setup.cfg`. If a contributor don't use `pre-commit`, we can just do the checks via Github Actions once they submit a PR.
> 
> Same case can be made for `isort` and `black`, but it is harder to decide since they are used as formatters. Right now `pre-commit run` runs them with `--check` so they just check if the code meets their specifications before a commit is made.

My understanding was that if they are in the dev dependencies (which is the case) then people with the dev setup will have them... and yes, I would also expect the regular actions CI to catch errors if people don't have the hooks installed.

However I would add the line about the pre-commit set up in the "Dev setup" section of the readme

File: .pre-commit-config.yaml (hunk 1 above)
Comment (alexnguyenn) [reply .pre-commit-config.yaml:36]: No need since we already specified it in `pyproject.toml`

File: .pre-commit-config.yaml (hunk 2 above)
Comment (alexnguyenn) [reply .pre-commit-config.yaml:26]: same reason above

**Comment by alexnguyenn** [comment]
Right now my understanding is `pre-commit` clone the tools you use from the repos you provide in `.pre-commit-config.yaml`, not from pypi. So there is no reason for contributors who are working with `pre-commit` to install `develop` dependencies as it will be redundant.

**Comment by gwwatkin** [comment]
> Right now my understanding is `pre-commit` clone the tools you use from the repos you provide in `.pre-commit-config.yaml`, not from pypi. So there is no reason for contributors who are working with `pre-commit` to install `develop` dependencies as it will be redundant.

Yes I noticed that when setting up ... then do we even really need the dev dependencies on these 4 pre commit hook packages?

**Review by gwwatkin — APPROVED** [review]
Looks good :+1: Added the set up instructions to the README

**Comment by alexnguyenn** [comment]
> Yes I noticed that when setting up ... then do we even really need the dev dependencies on these 4 pre commit hook packages?

Yeah that is my gripe. For people with `pre-commit` is redundant, but leaving those dependencies out for people without `pre-commit` means that they might not be aware that we are using those tools in development 😅 

I see you have been adding the contribution and set-up guide for `README.md`. We might want to move that under `CONTRIBUTION.md` for #117

**Comment by gwwatkin** [comment]
> Yeah that is my gripe. For people with `pre-commit` is redundant, but leaving those dependencies out for people without `pre-commit` means that they might not be aware that we are using those tools in development 😅 

:+1:  good thinking ... Also they are such light packages it doesn't really matter.

> I see you have been adding the contribution and set-up guide for `README.md`. We might want to move that under `CONTRIBUTION.md` for #117

:ok_hand:  left a [note on the issue](https://github.com/latticesurgery-com/lattice-surgery-compiler/issues/117#issuecomment-964685928), for when we set it up properly.

**Comment by gwwatkin** [comment]
I think we are good to merge

**Pull request merged**